generator/internal/stub_factory_generator.cc

```diff
@@ -73,22 +77,53 @@ Status StubFactoryGenerator::GenerateCc() {
   // clang-format on
 
   // includes
-  CcLocalIncludes({vars("stub_factory_header_path"),
-                   vars("logging_header_path"), vars("metadata_header_path"),
-                   vars("stub_header_path"), "google/cloud/log.h"});
+  CcLocalIncludes(
+      {vars("stub_factory_header_path"), vars("logging_header_path"),
+       vars("metadata_header_path"), vars("stub_header_path"),
+       "google/cloud/internal/algorithm.h",
+       "google/cloud/internal/common_options.h",
+       "google/cloud/internal/getenv.h", "google/cloud/internal/grpc_options.h",
+       "google/cloud/internal/options.h",
+       "google/cloud/internal/user_agent_prefix.h", "google/cloud/log.h"});
   CcSystemIncludes({"memory"});
   CcPrint("\n");
 
   auto result = CcOpenNamespaces(NamespaceType::kInternal);
   if (!result.ok()) return result;
 
+  CcPrint(  // clang-format off
+    "internal::Options Resolve$service_name$Options(internal::Options options) {\n"
+    "  if (!options.has<internal::EndpointOption>()) {\n"
+    "    auto env = internal::GetEnv(\"$service_endpoint_env_var$\");\n"
+    "    options.set<internal::EndpointOption>(env ? *env : \"$service_endpoint$\");\n"
+    "  }\n"
+    "  if (!options.has<internal::GrpcCredentialOption>()) {\n"
+    "    options.set<internal::GrpcCredentialOption>(grpc::GoogleDefaultCredentials());\n"
+    "  }\n"
+    "  if (!options.has<internal::GrpcBackgroundThreadsFactoryOption>()) {\n"
+    "    options.set<internal::GrpcBackgroundThreadsFactoryOption>(\n"
+    "        internal::DefaultBackgroundThreadsFactory);\n"
+    "  }\n"
+    "  if (!options.has<internal::GrpcNumChannelsOption>()) {\n"
+    "    options.set<internal::GrpcNumChannelsOption>(4);\n"
```

Comment:
This value of 4 here may be spanner specific. I'm not sure what a good default value is for other services. Not sure why 4 is better than 1 or N. You may want to start a thread w/ the team to see if anyone has any thoughts on this question.